src/ServiceProviderClient.php

```diff
@@ -302,11 +302,25 @@ public function authenticate($code, $redirectUri = null)
         // Decode the JWS token from inside the JWE token
         $jws = JOSE_JWT::decode($jwe->plain_text);
 
-        // Verify the JWS token
+        $header = $jws->header;  
+        $kid = $header['kid'];        
+
+        // Calculate cache experity time based on creation of cache file. In this example it is 10 minutes
+        // Of cource check that there is cache file at all 
+        if (file_exists("/tmp/isbcache.json")) {
+            $this->_isbSigningKeyRefreshTime = filemtime("/tmp/isbcache.json") + 10 * 60;
+        }
+
+        // Check if there is needs for cache update. In case cache does not exist keys are always fetched
         if (time() > $this->_isbSigningKeyRefreshTime) {
-            // Refresh ISB signing key
-            $public_key = $this->_getIsbSigningKey();
+            // Get new keys to cache (to local file) from JWKS endpoint
+            error_log("Old JWKS keys in cache, lets get new ones from JWKS");
+            $this->_getIsbSigningKeysToCache();
+        } else {
+            error_log("JWKS keys in cache, age less than 10 minutes");
         }
+
+        $public_key = $this->_getIsbSigningKeyFromCache($kid);
         $jws->verify($public_key);
```

Comment:
re-try would be nice. If getting the keys from local cache / verifying fails, it would be a good idea to refresh the local JWKS cache and then re-try verification. One-time re-try.

Reply:
Actually this is not necessary. Cache file does not exist immediately after the server in the container has ben started up. Cache will be created during the first authentication and cache refresh is based on cache file creation time. 

So, if someone delete the file in cache,  it will be automatically reloaded from JWSK end point. Logic is that $this->_isbSigningKeyRefreshTime default value is 0, and if timestamp cannot be obtained because no cache file, if (time() > $this->_isbSigningKeyRefreshTime) will be true and $this->_storeIsbSigningKeysToCache() will be called.